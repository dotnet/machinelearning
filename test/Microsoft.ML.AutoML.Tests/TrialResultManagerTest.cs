// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using FluentAssertions;
using Microsoft.ML.AutoML.CodeGen;
using Microsoft.ML.TestFramework;
using Xunit;
using Xunit.Abstractions;

namespace Microsoft.ML.AutoML.Test
{
    public class TrialResultManagerTest : BaseTestClass
    {
        public TrialResultManagerTest(ITestOutputHelper output)
            : base(output)
        {
        }

        [Fact(Skip = "failing part of ci test, need investigation")]
        [UseReporter(typeof(DiffReporter))]
        [UseApprovalSubdirectory("ApprovalTests")]
        public void CsvTrialResultManager_end_to_end_test()
        {
            var lgbmSearchSpace = new SearchSpace.SearchSpace<LgbmOption>();
            var tuner = new RandomSearchTuner(lgbmSearchSpace, 0);
            var trialResults = Enumerable.Range(0, 10)
                        .Select((i) =>
                        {
                            var trialSettings = new TrialSettings
                            {
                                TrialId = i,
                            };
                            var parameter = tuner.Propose(trialSettings);
                            trialSettings.Parameter = parameter;
                            return new TrialResult
                            {
                                TrialSettings = trialSettings,
                                DurationInMilliseconds = 10.123,
                                Loss = i * 0.99,
                                PeakCpu = i * 0.98,
                                PeakMemoryInMegaByte = i * 0.97,
                            };
                        });
            var tempFilePath = Path.Combine(OutDir, Path.GetRandomFileName() + ".txt");
            var csvTrialResultManager = new CsvTrialResultManager(tempFilePath, lgbmSearchSpace);

            // the tempFile is empty, so GetAllTrialResults should be 0;
            csvTrialResultManager.GetAllTrialResults().Count().Should().Be(0);

            // Add trialResults to csvTrialResultManager, the # of trials should be 10.
            foreach (var trialResult in trialResults)
            {
                csvTrialResultManager.AddOrUpdateTrialResult(trialResult);
            }
            csvTrialResultManager.GetAllTrialResults().Count().Should().Be(10);

            // if repeated trial added, csvTrialResultManager should update the existing trial.
            foreach (var trialResult in trialResults)
            {
                csvTrialResultManager.AddOrUpdateTrialResult(trialResult);
            }
            csvTrialResultManager.GetAllTrialResults().Count().Should().Be(10);

            // save as csv test
            csvTrialResultManager.Save();

            // reload test.
            csvTrialResultManager = new CsvTrialResultManager(tempFilePath, lgbmSearchSpace);
            csvTrialResultManager.GetAllTrialResults().Count().Should().Be(10);

            var fileContent = File.ReadAllText(tempFilePath);

            // replace line break to \r
            fileContent = fileContent.Replace(Environment.NewLine, "\r");
            Output.WriteLine(fileContent);
            File.Delete(tempFilePath);
            Approvals.Verify(fileContent);
        }
    }
}
