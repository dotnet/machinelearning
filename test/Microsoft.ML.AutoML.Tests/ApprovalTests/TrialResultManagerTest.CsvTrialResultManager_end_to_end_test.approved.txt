id,loss,durationInMilliseconds,peakCpu,peakMemoryInMegaByte,parameter_0,parameter_1,parameter_2,parameter_3,parameter_4,parameter_5,parameter_6,parameter_7,parameter_80,0.000,10.123,0.000,0.000,0.938,0.295,0.523,0.663,0.332,0.283,0.821,0.223,0.1871,0.990,10.123,0.980,0.970,0.069,0.625,0.693,0.630,0.927,0.217,0.929,0.067,0.9002,1.980,10.123,1.960,1.940,0.679,0.094,0.149,0.107,0.610,0.953,0.491,0.148,0.6783,2.970,10.123,2.940,2.910,0.570,0.225,0.618,0.005,0.361,0.850,0.062,0.282,0.7054,3.960,10.123,3.920,3.880,0.959,0.989,0.387,0.523,0.279,0.505,0.532,0.057,0.2815,4.950,10.123,4.900,4.850,0.496,0.846,0.613,0.104,0.742,0.401,0.933,0.131,0.7006,5.940,10.123,5.880,5.820,0.615,0.831,0.685,0.570,0.167,0.182,0.609,0.251,0.9977,6.930,10.123,6.860,6.790,0.569,0.063,0.264,0.843,0.567,0.646,0.516,0.970,0.0788,7.920,10.123,7.840,7.760,0.513,0.156,0.814,0.643,0.575,0.553,0.672,0.995,0.3029,8.910,10.123,8.820,8.730,0.064,0.466,0.483,0.216,0.066,0.228,0.151,0.447,0.451